### PR TITLE
Migrate styles to PostCSS

### PR DESCRIPTION
**Pull request by javivelasco**
This PR brings a complete migration of react-toolbox to [post-css](https://github.com/postcss/postcss) along with changes that are side effects of this migration. The theming philosophy behind this migration is basically the same as before. Each component has a configuration file where some variables are defined and used at time of generating styles. Those variables can be overridden to customize appearance. We are using:
- [postcss-import](https://github.com/postcss/postcss-import) allows to separate CSS in multiple files so we can do stuff like importing variables from one component into another.
- [postcss-mixins](https://github.com/postcss/postcss-mixins) allows to define and use mixins. They are not very used in react-toolbox but I've found them useful in the `Layout` component since some properties need to be repeated for different screen sizes.
- [postcss-next](http://cssnext.io/) the most important one. Includes multiple plugins to define custom variables, to use `calc`, etc.
- [postcss-each](https://github.com/outpunk/postcss-each) used in the `Layout` component to generate a bunch of classes that define sizes of the drawer element, for example.

With this implementation we don't need to [make tricks](https://github.com/react-toolbox/react-toolbox/blob/dev/docs/webpack.config.development.js#L52-L54) to prepend variables SASS files to react-toolbox stylesheets. We just need a [variables Javascript file](https://github.com/react-toolbox/react-toolbox/blob/css-next/docs/toolbox-variables.js) defining theming variables we want to override and [pass it to the custom properties plugin configuration](https://github.com/react-toolbox/react-toolbox/blob/css-next/docs/webpack.config.development.js#L65-L71). In case you are using sass variables from react-toolbox now, with the new version you'd need to change to postcss. Apart from that, if basically works the same way.

Of course this is a **huge** change because it implies a technology change and I think it should be released under `2.0.0`. The amazing news is that we can do awesome stuff with this change that I'd like to include in this PR before merging:
- [ ] Update MD documenting the usage of react-toolbox with PostCSS.
- [ ] Review and document the custom properties used for each component.
- [ ] More detailed documentation with specific info for Webpack and Theming.
- [ ] Redesign the docs.
- [ ] Include theming live customization in the docs playground.
- [ ] Create a live theme generator in the documentation.

Please leave your feedback in this PR!

### CODE REVIEW

**Comment by wtgtybhertgeghgtwtg** [comment]
>The amazing news is that we can do awesome stuff with this change that I'd like to include in this PR before merging:

Docs are one thing, but live theme editing in the playground seems like a bit much to hold this up.

**Comment by javivelasco** [comment]
Yeah, I agree with you but I also want to release the migration on NPM at the same time we can show the power and goodies it brings. The best way that comes to my mind is a live style editing and theme generator, that's why :)

**Comment by wtgtybhertgeghgtwtg** [comment]
You could cut a beta before releasing `2.0.0`.

**Comment by nelix** [comment]
This is great, I'm very excited for 2.0

File: docs/package.json [hunk 1]
```diff
@@ -39,7 +39,12 @@
     "html-webpack-plugin": "^2.16.1",
     "markdown-loader": "^0.1.7",
     "node-sass": "^3.7.0",
```
Comment (tbranyen) [inline docs/package.json]: This dependency (and the loader) can be dropped now, right?

**Comment by dev-guy** [comment]
I have not used react-toolbox. I'd prefer to start fresh with react-toolbox and postcss.

**Comment by ansumanshah** [comment]
@javivelasco , Exploring react toolbox for my project, just wanted to know when do you plan to merge this? :)

**Comment by javivelasco** [comment]
I'd like to update the documentation before merging adding a style editor and theme generator. Releasing officially in npm may take a while yet because of that. I'd like to do it at once to make noise. This branch is going to be updated with the latest changes but using postcss for styles anyway, so you can point here and start using it now. The best would be as @wtgtybhertgeghgtwtg said, to publish a beta with these changes before an official release.

**Comment by mmazzarolo** [comment]
@javivelasco I'm starting a new project and I plan to try give a try to this branch, do you have any hints on how to create a minimal webpack working setup?  
Thanks!

**Comment by ansumanshah** [comment]
@javivelasco 
I am using this branch in my project, so far so good :)
I think [this line](https://github.com/react-toolbox/react-toolbox/blob/css-next/lib/commons.css#L1) should be `@import '~normalize.css';`

**Comment by javivelasco** [comment]
Actually @ansumanshah, I think it's ok. This change is because now with PostCSS we are using `postcss-import` plugin and it takes cares of the imports. This means we are no longer importing CSS using webpack straight but through this plugin. Instead of using the tilde to tell webpack that import is coming from `node_modules` we can omit it because the plugin searchs in `node_modules` already.

@mmazzarolo Check the webpack configuration for the internal `spec` page [here](https://github.com/react-toolbox/react-toolbox/blob/css-next/webpack.config.development.js) and the docs [here](https://github.com/react-toolbox/react-toolbox/blob/css-next/docs/webpack.config.development.js). 

You have to take into account that to use a theme PostCSS takes the variables overrides from a Javascript file instead of prepending a SASS stylesheet like we used to. I think that's all! ✈️

**Comment by mmazzarolo** [comment]
Thank you @javivelasco, I asked that question because I'm having some trouble on importing and using `common.css` (webpack doesn't like at all the line `@import 'normalize.css'`)... I'm still investigating though (I have 0 skills on webpack).

**Comment by javivelasco** [comment]
Just recall that in this branch importing from CSS and from JS are resolved in different ways. If you are importing `normalize.css` from `commons.css` you should check the webpack configuration for `postcss-imports`. There are examples in the repository, it should work the same way.

**Comment by Reanmachine** [comment]
With the styling/theming switching to post-css, how do you want PRs involving style/theme changes to be submitted against the 1.x branch, basically it's increasing the transition work you need to do, and without the changes from `css-next` it'll be impossible to test any eye-balled implementation we could submit in pure post-css.

**Comment by javivelasco** [comment]
As long as the changes are not too heavy it's not very difficult to move them periodically to this branch. The final css files are pretty similar to SASS files so it's easy to move them. The most difficult work was to setup everything and to make a proof of concept of customization and live editing and it's done actually.

On my side there is no problem sending PRs to the dev branch without worrying about this transition :)

**Comment by ddeklerk** [comment]
> Include theming live customization in the docs playground.

If I'm reading this right, does it mean that the transition to PostCSS allows you to modify the theme during runtime?

I have next to zero experience with PostCSS, but I would love to know how something like that could be achieved. The project I am working on requires user made themes, so if this version of React-Toolbox makes that possible, that would be enormously helpful.

**Comment by javivelasco** [comment]
@ddeklerk you read it right. We can set up a theme using context with [react-css-themr](http://github.com/javivelasco/react-css-themr). The theme consists of some CSS stylesheets without variables interpolated that can be processed live with css-next and postcss-modules. The generated CSS can be injected live in the page and the generated module set into the context. This way we can generate themes live from a set of variables.

Here is a [proof of concept](https://twitter.com/javivelasco/status/765292988402630656) I've shared yesterday.

**Comment by razor-x** [comment]
Might I suggest separating out `normalize.css` so one has the option to load the common styles separately  from a global "reset" library? This potentially eases compatibility with other module loaders (since it's the only global library, issues with webpack expecting `~` might be bypassed).

For example, create a `main.css` (or `index.css`) file with


```css
@import 'normalize.css';
@import './common.css';
```

And then `common.css` is as before, but without the first line: 

```css
@import './colors.css';
@import './variables.css';

html {
  font-size: 62.5%;
}

body {
  font-family: Roboto, sans-serif;
// ...
```

**Comment by hisapy** [comment]
Hey guys,

I've made this branch work in my current setup although this is my first experience with React Toolbox. So far everything seems to be working ok. 

I'd like to know if there is better way to integrate _extracted css_ and the css provided by react-toolbox and if we really need to require all the `postcss-*` stuff

Here is my webpack.config.js

```javascript 
const webpack = require('webpack');
const path = require('path');
const ExtractTextPlugin = require("extract-text-webpack-plugin");

module.exports = {
  context: __dirname,
  entry: {
    app: ["./web/static/js/app.js"]
  },
  output: {
    path: "./priv/static",
    filename: "js/[name].js",
    chunkFilename: "js/[id].js"
  },

  postcss (webpackInstance) {
    return [
      require('postcss-import')({
        addDependencyTo: webpackInstance,
        root: __dirname,
        path: [
          path.join(__dirname, './web/static'),
          path.join(__dirname, './node_modules/react-toolbox/components')
        ]
      }),
      require('postcss-mixins')(),
      require('postcss-each')(),
      require('postcss-cssnext')({
        features: {
          customProperties: {
            variables: {
              'color-primary': 'var(--palette-indigo-500)',
              'color-primary-dark': 'var(--palette-indigo-700)',
              'color-primary-light': 'var(--palette-indigo-500)',
              'color-accent': 'var(--palette-pink-a200)',
              'color-accent-dark': 'var(--palette-pink-700)',
              'color-primary-contrast': 'var(--color-dark-contrast)',
              'color-accent-contrast': 'var(--color-dark-contrast)'
            }
          }
        }
      }),
      require('postcss-reporter')({ clearMessages: true })
    ];
  },

  plugins: [
    new ExtractTextPlugin("css/app.css", {
      allChunks: true // put css of all chunks in app.css
    }),
    new webpack.optimize.CommonsChunkPlugin('js/commons.chunk.js')
  ],

  module: {
    loaders: [{
      test: /\.jsx?$/,
      loader: 'babel',
      exclude: /node_modules/,
      query: {
        plugins: [
          path.resolve(__dirname, 'web', 'static', 'js', 'babelRelayPlugin'),
          'transform-runtime'
        ],
        presets: ['react', 'es2015', 'stage-0']
      }
    },{
      test: /\.css$/,
      // Extracting these css because I'm not sure how react-toolbox or postcss work
      include: /graphiql|web\/static\/css\/app\.css/,
      loader: ExtractTextPlugin.extract("style", "css")
    },{
      test: /\.css$/,
      exclude: /graphiql|web\/static\/css\/app\.css/,
      loader: "style-loader!css-loader?modules&importLoaders=1!postcss-loader"
    }]
  }
};
```

> As a side note, we just picked this library instead of going with material-ui because  the demo was performing better on my old iPad 2. We were doing our own components based on jQuery and Bootstrap 3 before. The only problem I noticed in the iPad is the scroll. It lacks momentum and motion.

**Comment by javivelasco** [comment]
It looks good! I think we may need only css-next because what we want you to generate is the variables. You shouldn't have to resolve imports, mixins etc, that remains internal. We'd need to change the build setup so instead of copying CSS files we can process everything but css-next

**Comment by razor-x** [comment]
I agree with @javivelasco. Requiring people to use postcss plugins before css module generation is not much different then requiring Sass as is done now. At that point you are are just using **post**css as an alternative **pre**processor.

Any  (post)CSS features still included in the package for the enduser that generate classes dynamically  (like imports) would need to be run by the user before generating the CSS module (just like Sass now).

Bundling via postcss can be done before packaging this lib to provide  CSS ready to be loaded as modules by the user. Then they can pass their final (modularized) stylesheet(s) though css-next and their other postcss plugs.

**Comment by mmazzarolo** [comment]
@javivelasco this brach rocks and it's working flawlessly for me (well, at least after the setup 🐙).  
Thank you! 👍

**Comment by javivelasco** [comment]
Thank you @mmazzarolo !!

**Comment by mmazzarolo** [comment]
I'd like to show the only two issues I'm having with this branch at the moment:
  
### Importing commons.css
I can't import `'react-toolbox/lib/commons.css'`, I'm having the following error:
```ERROR in ./~/react-toolbox/lib/commons.css
Module build failed: ModuleBuildError: Module build failed: Error: ENOENT: no such file or directory, open '/mypjpath/node_modules/react-toolbox/lib/normalize.css'```
I need to import it because otherwise I fall in [this issue](https://github.com/react-toolbox/react-toolbox/issues/347).   
Not a big problem though, I just copied `commons.css` and its css dependencies in a css file in my own project to make it work. It should be related to the webpack loaders I guess... 

### Declaring a global theme with cssnext
I just noticed that passing custom variables to the postcss-cssnext plugin doesn't override the default theme for me 🐒. 
Anyone else experiencing this issue?  

```javascript
// postcss-cssnext config
require('postcss-cssnext')({
  features: {
    customProperties: {
      variables: {
        'color-primary': 'var(--palette-teal-500)',
        'color-primary-dark': 'var(--palette-teal-700)',
        'color-primary-light': 'var(--palette-teal-500)',
        'color-accent': 'var(--palette-amber-a200)',
        'color-accent-dark': 'var(--palette-amber-700)',
        'color-primary-contrast': 'var(--color-dark-contrast)',
        'color-accent-contrast': 'var(--color-dark-contrast)'
      }
    }
  }
})
```

```javascript
// loaders
loaders: [
  {
    test: /\.(js|jsx)$/,
    include: srcPath,
    exclude: nodeModulesPath,
    loaders: ['react-hot', 'babel']
  }, {
    test: /\.css$/,
    include: srcPath,
    exclude: reactToolboxPath,
    loader: 'style-loader!css-loader?localIdentName=[local]__[path][name]__[hash:base64:5]&modules&importLoaders=1&sourceMap!postcss-loader'
  }, {
    test: /\.css$/,
    include: reactToolboxPath,
    exlude: nodeModulesPath,
    loader: ExtractTextPlugin.extract('style', 'css?sourceMap&modules&importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]!postcss')
  }, {
    test: /\.(jpg|png|gif)$/,
    loaders: ['file-loader']
  }
]
```

P.S.: The two issues don't seems to be related (I'm still having the second issue even if I don't import my own commons.css file).  
Thanks!

**Comment by javivelasco** [comment]
Hi @mmazzarolo !! I've tried the cssnext branch from the example repository and it was working nicely. Maybe webpack is not able to find the postcss configuration you are specifying or overriding it some way. I've been trying this to transform css live and in the example and it works like a charm.

About the first issue you probably have to configure postcss-include plugin passing down the webpack configuration so it can discover packages located in css modules. Check the configuration included in the documentation and the spec in this same branch and let me know!

**Comment by tbranyen** [comment]
@mmazzarolo have you confirmed that your npm installed node_modules contains the right version of react-toolbox? I'd confirm the CSS files exist.

**Comment by mmazzarolo** [comment]
> About the first issue you probably have to configure postcss-include plugin passing down the webpack configuration so it can discover packages located in css modules. Check the configuration included in the documentation and the spec in this same branch and let me know!  

Yup, I did try just like you suggested in your first comment but it didn't work... can't remember the the reason though.  
I'll take a look at it again then, I just wanted to know if these were known issues... or just the consequences of my incompetence in webpack as I suspected 👯 

> have you confirmed that your npm installed node_modules contains the right version of react-toolbox? I'd confirm the CSS files exist.

Yes, it seems to be the right version... everything else works correctly for me. 

Thanks!

**Comment by javivelasco** [comment]
FYI as soon as we get the first three checkboxes checked in the PR first comment we are releasing. Waiting to release this branch until we redesign the docs and improve the playground may take some time and it shouldn't be a blocker. The noise should come with the new playground release to show the full potential of this migration so IMO the best would be to release this branch quietly and then, with new docs and a new playground, we make some noise :)

Opinions?

**Comment by Reanmachine** [comment]
@javivelasco If I was interested in helping move along the 2.0 release, what areas would be most useful for me to help in, as far as getting those first 3 checkboxes done.

**Comment by javivelasco** [comment]
Thanks @Reanmachine ! 

The first thing would be to figure out an structure for the documentation with strategical tips and develop a big those docs in markdown format. They should be accesible first through the repository. 

Then we can go through each component trying to minimize the priority of class selectors. For example, the `Button` component now has a `:not([disabled])` modifier that is annoying to override. The more flat the css file the better. Then we should review CSS custom properties that are likely to be modified and namespace them in case they are not (most of them already are). This allows to identify which component owns each property from the future `theme.js` file. Finally, we should create a new section for each component documenting css custom properties.

Once that's all done we can review again what's next. The reason is that I think it could be cool to change a little the look and feel of the documentation allowing to show how the components look without the need of loading them in the playground. I'm getting help from some designers so by that time maybe I have some insights on how to improve the docs site instead of changing stuff beforehand. If anybody wants to help ping me through our discord channel!

**Comment by mmazzarolo** [comment]
If anyone is interested I just created a [minimal working boilerplate using this branch](https://github.com/mmazzarolo/react-toolbox-postcss-boilerplate).  
I also solved my previous issues: `postcss-import` must be required **BEFORE** `postcss-next`. My bad, it should have been obvious.  
With the boilerplate above you can also see that the only required setup for make this branch work is using `postcss-import` and `postcss-next`: 
```javascript
postcss (webpackInstance) {
    return [
      require('postcss-import')({
        addDependencyTo: webpackInstance,
        root: rootPath,
        path: [
          path.join(reactToolboxPath)
        ]
      }),
      require('postcss-cssnext')({
        features: {
          customProperties: {
            variables: {
              'color-primary': 'var(--palette-teal-500)',
              'color-primary-dark': 'var(--palette-teal-700)',
              'color-primary-light': 'var(--palette-teal-500)',
              'color-accent': 'var(--palette-amber-a200)',
              'color-accent-dark': 'var(--palette-amber-700)',
              'color-primary-contrast': 'var(--color-dark-contrast)',
              'color-accent-contrast': 'var(--color-dark-contrast)'
            }
          }
        }
      })
    ]
  }
```

Bye!

**Comment by javivelasco** [comment]
I wanted to add that when we release this new version, the build will include every css processed except for the css-next so postcss-include will not be necessary!

**Comment by razor-x** [comment]
Any thoughts on separating out normalize.css per my earlier comment (I think it got buried): https://github.com/react-toolbox/react-toolbox/pull/666#issuecomment-240544091

**Comment by acornejo** [comment]
Tried using the beta today. Configuring postcss-import to work with react-toolbox was annoying since I had to fiddle with relative paths and add react-toolbox to the include list. Is there any chance we can simplify this a bit to make adding this to a new project a bit more painless?

After getting past postcss-import issues I was greeted with the following:

ERROR in ./~/css-loader!./~/postcss-loader!./~/react-toolbox/lib/button/theme.css
Module build failed: Error: composition is only allowed when selector is single :local class name not in ".raised", ".raised" is weird
 @ ./~/react-toolbox/lib/button/theme.css 4:14-104 13:2-17:4 13:2-17:4 14:20-110

**Comment by javivelasco** [comment]
Of course @acornejo ! This was released just for people using the branch directly but I'm releasing tomorrow a new beta with imports and stuff already processed so you just would need to configure postcss next to customize variables. No need for anything else judt custom properties :)

**Comment by wtgtybhertgeghgtwtg** [comment]
Was there a problem with the preprocessed build?

**Comment by javivelasco** [comment]
Not at all, I just didn't have time enough to finish and push it! Almost done locally

**Comment by jjstrat3** [comment]
@javivelasco I really appreciate all your work and efforts towards this refactor!  I see there's some conflicts.  When you're able to resolve those, will this branch be merged into a new beta release?

**Comment by javivelasco** [comment]
Sure, didn't have time enough and I'm trying out another approach. The only thing remaining here is to write some documentation and release the build system. I'm trying to open an issue detailing the progress tomorrow to organize this. Otherwise this would take forever, help is appreciated!

**Comment by razor-x** [comment]
I'm unable to get css-next to override the vars, but hopefully I can work out the issue.

This got me thinking though, with the current structure of `theme.css` there is no way to roll multiple custom versions of a component to use in a single app.

Say I want to make a smaller button I use in a few places. I'd like to create (and then pass in) a custom theme by making a css module like

```css
:root {
  --button-height: 25px;
}

@import 'react-toolbox/components/button/theme.css';
```

But since `react-toolbox/components/button/theme.css` always does

```css
@import '../colors.css';
@import '../variables.css';
@import './config.css';

/* ... actual button styles ... */
```

my custom value for `--button-height` always gets overridden.

We could solve this by moving the `/* ... actual button styles ... */` to `styles.css` and change `theme.css` to

```css
@import '../colors.css';
@import '../variables.css';
@import './config.css';
@import './styles.css';
```

Then one may make a custom button with, e.g.,

```css
@import 'react-toolbox/components/colors.css';
@import 'react-toolbox/components/variables.css';

:root {
  --unit: 0.5rem;
}

@import 'react-toolbox/components/button/config.css';

:root {
  --button-squared-icon-margin: 5px;
}

@import 'react-toolbox/components/button/theme.css';
```

I can probably script a find / replace and submit a PR if we want to support this.

**Comment by carlitux** [comment]
@razor-x I put this on webpack to configure postcss

```js
  postcss: function (webpackInstance) {
    return [
      postcssImport({
        addDependencyTo: webpackInstance,
        root: rootPath,
        path: [
          path.join(reactToolboxPath),
        ],
      }),
      require('postcss-mixins')(),
      require('postcss-each')(),
      postcssCssNext({
        features: {
          customProperties: {
            variables: require('../src/theme/variables.js'),
          }
        }
      }),
      // precss,
      // autoprefixer
    ];
  },
```

and the variables.js

```js
module.exports = {
  'color-divider': 'var(--palette-grey-200)',
  'color-background': 'var(--color-white)',
  'color-text': 'var(--palette-grey-900)',
  'color-text-secondary': 'var(--palette-grey-600)',

  'color-primary': 'var(--palette-red-500)',
  'color-primary-dark': 'var(--palette-red-700)',
  'color-accent': 'var(--palette-pink-a200)',
  'color-accent-dark': 'var(--palette-pink-700)',
  'color-primary-contrast': 'var(--color-dark-contrast)',
  'color-accent-contrast': 'var(--color-dark-contrast)',

  unit: '1rem',

  'preferred-font': '"Roboto", "Helvetica", "Arial", "ans-serif"',
  'font-size': 'calc(1.6 * var(--unit))',
  'font-size-tiny': 'calc(1.2 * var(--unit))',
  'font-size-small': 'calc(1.4 * var(--unit))',
  'font-size-normal': 'var(--font-size)',
  'font-size-big': 'calc(1.8 * var(--unit))',
  'font-weight-thin': '300',
  'font-weight-normal': '400',
  'font-weight-semi-bold': '500',
  'font-weight-bold': '700',

  'shadow-2p':
    '0 2px 2px 0 rgba(0, 0, 0, 0.14),' +
    '0 3px 1px -2px rgba(0, 0, 0, 0.2),' +
    '0 1px 5px 0 rgba(0, 0, 0, 0.12);',

  'shadow-3p':
    '0 3px 4px 0 rgba(0, 0, 0, 0.14),' +
    '0 3px 3px -2px rgba(0, 0, 0, 0.2),' +
    '0 1px 8px 0 rgba(0, 0, 0, 0.12);',

  'shadow-4p':
    '0 4px 5px 0 rgba(0, 0, 0, 0.14),' +
    '0 1px 10px 0 rgba(0, 0, 0, 0.12),' +
    '0 2px 4px -1px rgba(0, 0, 0, 0.2);',

  'shadow-6p':
    '0 6px 10px 0 rgba(0, 0, 0, 0.14),' +
    '0 1px 18px 0 rgba(0, 0, 0, 0.12),' +
    '0 3px 5px -1px rgba(0, 0, 0, 0.2);',

  'shadow-8p':
    '0 8px 10px 1px rgba(0, 0, 0, 0.14),' +
    '0 3px 14px 2px rgba(0, 0, 0, 0.12),' +
    '0 5px 5px -3px rgba(0, 0, 0, 0.2);',

  'shadow-16p':
    '0 16px 24px 2px rgba(0, 0, 0, 0.14),' +
    '0 6px 30px 5px rgba(0, 0, 0, 0.12),' +
    '0 8px 10px -5px rgba(0, 0, 0, 0.2);',

  'shadow-key-umbra-opacity': 0.2,
  'shadow-key-penumbra-opacity': 0.14,
  'shadow-ambient-shadow-opacity': 0.12,

  'zdepth-shadow-1':
    '0 1px 6px rgba(0, 0, 0, 0.12), 0 1px 4px rgba(0, 0, 0, 0.24)',
  'zdepth-shadow-2':
    '0 3px 10px rgba(0, 0, 0, 0.16), 0 3px 10px rgba(0, 0, 0, 0.23)',
  'zdepth-shadow-3':
    '0 10px 30px rgba(0, 0, 0, 0.19), 0 6px 10px rgba(0, 0, 0, 0.23)',
  'zdepth-shadow-4':
    '0 14px 45px rgba(0, 0, 0, 0.25), 0 10px 18px rgba(0, 0, 0, 0.22)',
  'depth-shadow-5':
    '0 19px 60px rgba(0, 0, 0, 0.3), 0 15px 20px rgba(0, 0, 0, 0.22)',

  'animation-duration': '0.35s',
  'animation-delay': 'calc(var(--animation-duration) / 5)',
  'animation-curve-fast-out-slow-in': 'cubic-bezier(0.4, 0, 0.2, 1)',
  'animation-curve-linear-out-slow-in': 'cubic-bezier(0, 0, 0.2, 1)',
  'animation-curve-fast-out-linear-in': 'cubic-bezier(0.4, 0, 1, 1)',
  'animation-curve-default': 'var(--animation-curve-fast-out-slow-in)',

  'z-index-highest': 300,
  'z-index-higher': 200,
  'z-index-high': 100,
  'z-index-normal': 1,
  'z-index-low': -100,
  'z-index-lower': -200,
};
```

in this way works for me.

**Comment by patrick-jones** [comment]
@javivelasco , I've done a quick review of the new layout and have a few bugs to report. Do you want to discuss it here or on separate issues?

**Comment by javivelasco** [comment]
Of course! Feel free to open a issue with them :)

**Comment by mgcrea** [comment]
Really excited to drop sass and use postcss! Have you noticed any performance change vs saas regarding generation time? Any ETA for the "official" beta release?

**Comment by borela** [comment]
When you talk about live editing the variables, does that mean that I won't need to rebuild the project? Like having a theme switch system without having css for each variation? Right now, every time I change a SCSS file, webpack HMR does the job of recompiling and updating the page but this way the final CSS is fixed.

Will it be possible to use context theming like before? Currently I have a folder with the SCSS files for the components I need to change.

theme

**Comment by javivelasco** [comment]
@borela I mean the playground will have multiple files to edit including CSS Modules so it's going to be possible to change a CSS file or a configuration variable in the playground and it's going to be updated live. But I'm always referring to the playground. In your local of course it's possible but it depends on your configuration :)

**Comment by kjg531** [comment]
Where would you suggest someone go to start the learning curve in order to migrate to this in the future?

**Comment by javivelasco** [comment]
@kjg531 I'd go straight for this. Although it is not officially released it is not because it's not stable but because there are a couple of things regarding tooling that are important and still pending

**Comment by pekkah** [comment]
So how do I get started with this branch? Is there a beta, alpha, something released on npm?

**Comment by maitriyogin** [comment]
Hiya, can I use this branch with creat-react-app out of the box, without ejecting?
/Stephen

**Comment by rubenmoya** [comment]
No @maitriyogin, you have to eject to be able to configure PostCSS

**Comment by ss74** [comment]
No

**Comment by maitriyogin** [comment]
Ok, so a double negative... I'll try both and see. I know that cra has support for postcss.  I'll fiddle and let you know.
/stephen

**Comment by ss74** [comment]
Ok TQ

**Comment by javivelasco** [comment]
@maitriyogin There will be a build system. I'm working on it this week to merge this branch and keep doing more stuff. Otherwise this will be kept forever

**Comment by kjg531** [comment]
How do you suggest I configure this with postCSS with ssr?

**Comment by kjg531** [comment]
PostCSS and SSR*

**Comment by retorquere** [comment]
I'm looking to give this a go because I share the need expressed in #963. How do I get started on this branch? I've tried @mmazzarolo 's bolierplate but that gives me

```
ERROR in ./src/index.js
Module not found: Error: Cannot resolve module 'react/lib/ReactMount' in /Users/emile/heroku/react-toolbox-postcss-boilerplate/src
 @ ./src/index.js 1:340-371

ERROR in ./src/index.js
Module not found: Error: Cannot resolve module 'react-toolbox/lib/commons.css' in /Users/emile/heroku/react-toolbox-postcss-boilerplate/src
 @ ./src/index.js 15:0-40

ERROR in ./src/App.js
Module not found: Error: Cannot resolve module 'react/lib/ReactMount' in /Users/emile/heroku/react-toolbox-postcss-boilerplate/src
 @ ./src/App.js 1:340-371

ERROR in ./src/App.js
Module not found: Error: Cannot resolve module 'react-toolbox/lib/card' in /Users/emile/heroku/react-toolbox-postcss-boilerplate/src
 @ ./src/App.js 17:12-45

ERROR in ./src/App.js
Module not found: Error: Cannot resolve module 'react-toolbox/lib/app_bar' in /Users/emile/heroku/react-toolbox-postcss-boilerplate/src
 @ ./src/App.js 19:15-51

ERROR in ./src/App.js
Module not found: Error: Cannot resolve module 'react-toolbox/lib/button' in /Users/emile/heroku/react-toolbox-postcss-boilerplate/src
 @ ./src/App.js 23:14-49

ERROR in ./src/App.js
Module not found: Error: Cannot resolve module 'react-toolbox' in /Users/emile/heroku/react-toolbox-postcss-boilerplate/src
 @ ./src/App.js 15:20-44
```

on `npm start`

**Comment by javivelasco** [comment]
[Here])(github.com/react-toolbox/react-toolbox-themr) is the build system. I forgot to mention here!

**Comment by aislanmaia** [comment]
Okay, I trying to use this beta but right now I'm getting problems like:

ERROR in ./app/components/students/StudentForm/ResponsibleFormSection.js
Module not found: Error: Cannot resolve module 'react-toolbox/lib/input' in /home/aislan/Projects/Phoenix/AppNaEscola/app-web-frontend/app/components/students/StudentForm
 @ ./app/components/students/StudentForm/ResponsibleFormSection.js 23:13-47

ERROR in ./app/components/students/StudentForm/ResponsibleFormSection.js
Module not found: Error: Cannot resolve module 'react-toolbox/lib/dropdown' in /home/aislan/Projects/Phoenix/AppNaEscola/app-web-frontend/app/components/students/StudentForm
 @ ./app/components/students/StudentForm/ResponsibleFormSection.js 27:16-53

ERROR in ./app/components/students/StudentConfirmDialog/index.js
Module not found: Error: Cannot resolve module 'react-toolbox/lib/dialog' in /home/aislan/Projects/Phoenix/AppNaEscola/app-web-frontend/app/components/students/StudentConfirmDialog
 @ ./app/components/students/StudentConfirmDialog/index.js 13:14-49

Before the components was getting compiled. 

This is my webpack config dev: 

[webpack.config.js.txt](https://github.com/react-toolbox/react-toolbox/files/728405/webpack.config.js.txt)

**Comment by javivelasco** [comment]
@aislanmaia Your configuration needs to support CSS modules with postcss for react-toolbox. I see this line here:

```js
loader: ExtractTextPlugin.extract('style', 'css?sourceMap&modules&importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]!postcss!sass'),
```

I guess that's the loader that was being used for react-toolbox. You just have to remove `!sass` from there and to add `css-next` in the configuration for `postcss`. Right now the config is:

```js
postcss: function () {
   return [autoprefixer];
}
```

I recommend to use just css-next. Check out the configuration examples in the root of the repository and the /docs folder. Also, you may use that loader config just for react-toolbox because right now there are two different loaders with the same regex!

**Comment by aldensc** [comment]
Any timeline on the official 2.0 release?  Downloaded the npm install and it is still in beta.  I see the comment on fixing up the documentation and playground, but that comment was in Jul 2016.  Personally, I would be good with some documentation and have the playground come later.  I've had to dig through the code to find out how to use certain components.  2.0 looks like it may fulfill our needs, but would not be allowed to use a beta.

**Pull request merged**